pkg/provision/cleanup.go

```diff
@@ -0,0 +1,230 @@
+						}
+						toDelete[filepath.Base(mnt.Source)] = struct{}{}
+					} else {
+						toSave[filepath.Base(mnt.Source)] = struct{}{}
+					}
+				}
+			}
+		}
+	}
+
+	ctx := context.TODO()
+	utils := filesystem.NewUtils()
+	pools, err := ioutil.ReadDir("/mnt")
+	if err != nil {
+		return errors.Wrap(err, "failed to list available pools")
+	}
+
+	for _, pool := range pools {
+		if !pool.IsDir() {
+			continue
+		}
+		path := filepath.Join("/mnt", pool.Name())
+		subvols, err := utils.SubvolumeList(ctx, path)
+		if err != nil {
+			// A Pool might not be mounted here, skip it
+			log.Error().Err(err).Msgf("failed to list available subvolumes in: %s", path)
+			continue
+		}
+
+		qgroups, err := utils.QGroupList(ctx, path)
+		if err != nil {
+			return errors.Wrapf(err, "failed to list available cgroups in: %s", path)
+		}
+
+		for _, subvol := range subvols {
+			qgroup, ok := qgroups[fmt.Sprintf("0/%d", subvol.ID)]
+			if !ok {
+				log.Info().Msgf("skipping volume '%s' has no assigned quota", subvol.Path)
+				continue
+			}
+
+			// now, is this subvol in one of the toDelete ?
+			if _, ok := toDelete[filepath.Base(subvol.Path)]; ok {
+				// delete the subvolume
+				log.Info().Msgf("deleting subvolume %s", subvol.Path)
+				if err := utils.SubvolumeRemove(ctx, filepath.Join(path, subvol.Path)); err != nil {
+					log.Err(err).Msgf("failed to delete subvol '%s'", subvol.Path)
+				}
+				if err := utils.QGroupDestroy(ctx, qgroup.ID, path); err != nil {
+					log.Err(err).Msgf("failed to delete qgroup: '%s'", qgroup.ID)
+				}
+				continue
+			}
+
+			// we only handle volumes that are 256MiB or 10MiB
+			if qgroup.MaxRfer != 268435456 && qgroup.MaxRfer != 10485760 {
+				// if the subvolume is a zdb and has 0 maxRfer, don't skip here. It might need to be deleted
+				if !strings.HasPrefix(subvol.Path, "zdb") && qgroup.MaxRfer != 0 {
+					log.Info().Msgf("skipping volume '%s' is of size: %d", subvol.Path, qgroup.MaxRfer)
+					continue
+				}
+			}
+
+			// now, is this subvol in one of the toSave ?
+			if _, ok := toSave[filepath.Base(subvol.Path)]; ok {
+				log.Info().Msgf("skipping volume '%s' is used", subvol.Path)
+				continue
+			}
+
+			// delete the subvolume
+			log.Info().Msgf("deleting subvolume %s", subvol.Path)
+			if err := utils.SubvolumeRemove(ctx, filepath.Join(path, subvol.Path)); err != nil {
+				log.Err(err).Msgf("failed to delete subvol '%s'", subvol.Path)
+			}
+			if err := utils.QGroupDestroy(ctx, qgroup.ID, path); err != nil {
+				log.Err(err).Msgf("failed to delete qgroup: '%s'", qgroup.ID)
+			}
+		}
+	}
+
+	return nil
+}
+
+func deleteZdbContainer(containerID pkg.ContainerID) error {
+	zbusCl, err := zbus.NewRedisClient("unix:///var/run/redis.sock")
+	if err != nil {
+		log.Fatal().Err(err).Msg("fail to connect to message broker server")
+	}
+
+	container := stubs.NewContainerModuleStub(zbusCl)
+	flist := stubs.NewFlisterStub(zbusCl)
+
+	info, err := container.Inspect("zdb", containerID)
```

Comment:
May be you should add an `Exists` method to the container API. because inspect can return other errors as well.

Reply:
Yeah that might be nice, but this should not be added in this PR